Documentation/contributing/development/statedb.rst

```diff
@@ -0,0 +1,443 @@
+
+* Write benchmarks for your table to understand the cost of the indexing and storage use. See ``benchmarks_test.go`` in ``cilium/statedb`` for examples.
+
+* If the object is small (<100 bytes) prefer storing it by value instead of by reference, e.g. ``Table[MyObject]`` instead of ``Table[*MyObject]``. This reduces memory fragmentation and makes it safer to use since the fields can't be accidentally mutated (anything inside that's by reference of course can be mutated accidentally). Note though that each index will store a separate copy of the object. Measure if needed.
+
+With that out of the way, let's get concrete with a code example of a simple
+table and a controller that populates it:
+
+.. literalinclude:: ../../../contrib/examples/statedb/example.go
+   :language: go
+
+To understand how the table defined by our example module can be consumed, we can
+construct a small mini-application:
+
+.. literalinclude:: ../../../contrib/examples/statedb/main.go
+   :language: go
+
+You can find and run the above examples in ``contrib/examples/statedb``:
+
+.. code-block:: shell
+
+  contrib/examples/statedb$ go run .
+
+
+Pitfalls
+^^^^^^^^
+
+Here are some common mistakes to be aware of:
+
+* Object is mutated after insertion to database. Since StateDB queries do not return copies, all readers will see the modifications.
+
+* Object (stored by reference, e.g. ``*T``) returned from a query is mutated and then inserted. StateDB will catch this and panic. Objects stored by reference must be (shallow) cloned before mutating.
+
+* Query is made with ReadTxn and results are used in a WriteTxn. The results may have changed between the ReadTxn and WriteTxn! If you want optimistic concurrency control, then use CompareAndSwap in the write transaction.
+
+cilium-dbg statedb
+^^^^^^^^^^^^^^^^^^
+
+StateDB implements a simple HTTP API for querying tables from a separate process.
+The objects are marshalled and unmarshalled using the normal Go JSON facilities,
+so we can effectively get copies of the objects into a separate process using the
+HTTP API which we can then render or process.
+
+The ``cilium-dbg statedb <command>`` provides a table-based view into the StateDB
+tables running in the agent in a style similar to "kubectl get". Watching the
+tables for changes is also supported with the ``--watch`` flag.
+
+To get your newly defined table to be part of this, there's only two things you
+need to do: implement the *TableRow* and *TableHeader* methods for the object (see above)
+and add a one-liner to ``cilium-dbg/cmd/statedb.go``:
+
+.. code-block:: go
+
+	func init() {
+		...
+		StatedbCmd.AddCommand(
+			...
+			statedbTableCommand[example.Example](example.TableName),
+		)
+	}
+
+We can now dump the contents of our example table:
+
+.. code-block:: shell
```

Comment:
```suggestion
.. code-block:: shell-session
```

Although the pound sign will mess up formatting a bit (it will be considered as a prompt symbol).